### PR TITLE
Simplify Percy Breakpoint/Theme snapshots

### PR DESCRIPTION
**Pull request by jmuzina**
## Done

- Light examples are snapshotted responsively
- Dark/paper examples are snapshotted on desktop only
- Slightly simplified / cleaned up snapshots JS code by adding some constants to reduce literal duplication

Fixes [WD-13804](https://warthogs.atlassian.net/browse/WD-13804)

## QA

- View [test build](https://percy.io/bb49709b/test-vanilla-gha-migration/builds/35631002/changed/1946954265?browser=chrome&browser_ids=59&group_snapshots_by=similar_diff&subcategories=approved&viewLayout=overlay&viewMode=new&width=1280&widths=375%2C800%2C1280) and verify snapshots are captured as expected
  - Pay special attention to the relationship between theme and breakpoints
    - All snapshots are captured at 1280px (desktop/large size)
    - Light-theme examples are also captured at 375px (mobile/small size)
    - Examples that include "responsive" in their name or embed such an example are also captured at 800px (tablet/medium size)

### Check if PR is ready for release

If this PR contains Vanilla SCSS code changes, it should contain the following changes to make sure it's ready for the release:

- [x] PR should have one of the following labels to automatically categorise it in release notes:
  - `Feature 🎁`, `Breaking Change 💣`, `Bug 🐛`, `Documentation 📝`, `Maintenance 🔨`.
- [ ] Vanilla version in `package.json` should be updated relative to the [most recent release](https://github.com/canonical/vanilla-framework/releases/latest), following semver convention:
  - if CSS class names are not changed it can be bugfix release (x.x.**X**)
  - if CSS class names are changed/added/removed it should be minor version (x.**X**.0)
  - see the [wiki for more details](https://github.com/canonical/vanilla-framework/wiki/Release-process#pre-release-tasks)
- [ ] Any changes to component class names (new patterns, variants, removed or added features) should be listed on the [what's new page](https://github.com/canonical/vanilla-framework/blob/main/releases.yml).

## Screenshots
image

[WD-13804]: https://warthogs.atlassian.net/browse/WD-13804?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by jmuzina** [comment]
This is progressing nicely, but I've noticed that several snapshots are missing. It looks like some error is being thrown for snapshots that are not combined. Error message is::
```
[percy:core] Error: Mysql2::Error: Duplicate entry '35527075-2565' for key 'index_test_case_executions_on_build_id_and_test_case_id'
    at handleFinished (file:///home/runner/work/vanilla-framework/vanilla-framework/node_modules/@percy/client/dist/utils.js:200:17)
    at IncomingMessage.<anonymous> (file:///home/runner/work/vanilla-framework/vanilla-framework/node_modules/@percy/client/dist/utils.js:216:27)
    at IncomingMessage.emit (node:events:531:35)
    at endReadableNT (node:internal/streams/readable:1696:12)
    at process.processTicksAndRejections (node:internal/process/task_queues:82:21) (0ms)
```
Sent a message to Browserstack for help diagnosing.

**Comment by jmuzina** [comment]
Still awaiting response from Browserstack, so I separated the test case grouping into its own [task](https://warthogs.atlassian.net/browse/WD-13805). This PR will just implement the simplified breakpoint/theming grouping

**Comment by jmuzina** [comment]
@pastelcyborg 
> I must be missing some context here. Why are we not capturing screenshots of each theme at every breakpoint?

We currently take one snapshot per theme at desktop size, and one snapshot per additional breakpoint on default theme (with no theme query param specified). So for a given example (say, tooltips combined, you get the following snapshots)

| Query Param      | Widths      |
| ------------- | ------------- |
| ?theme=light | [1280] |
| ?theme=paper | [1280] |
| ?theme=dark | [1280] |
| N/A (defaults to light) | [375] |

This means you have to review four different snapshot groupings... multiplied by two for standalone.

This change will take one snapshot per breakpoint on light theme, and an additional desktop snapshot per color theme. So that would change to:

| Query Param      | Widths      |
| ------------- | ------------- |
| ?theme=light | [375, 1280] |
| ?theme=paper | [1280] |
| ?theme=dark | [1280] |

The amount of snapshots is the same, but Percy makes it easier to find the light responsive snapshot by grouping them together like this:

image

Rather than having to scroll through and select the other snapshot in the snapshots list, like you currently do:

image

We will have 6 snapshot groupings per example, rather than 8.
image

**Comment by jmuzina** [comment]
> Why are we not capturing screenshots of each theme at every breakpoint?

We intentionally don't capture at *every* breakpoint/theme combination. We assume that snapshot responsive content handling doesn't vary between themes at the same breakpoint, so we only capture the smaller breakpoints on desktop.

**Comment by jmuzina** [comment]
> Just to make sure I understand this: we don't currently take mobile screenshots for dark and paper theme? I didn't realize that; that's not optimal. I sort of understand the logic, but I'd probably suggest we take screenshots of all themes across all breakpoints eventually.

@pastelcyborg  This comes from this [MM message](https://chat.canonical.com/canonical/pl/wh61omg7tibu9nyxy5g6ugwr9w) from @bartaz  - it was done as a way of reducing test volume I believe. If we captured in all themes and breakpoints, we'd expect to see test volume increase by around 3x. However, I do agree that it's optimal to snapshot every breakpoint in every theme for thoroughness' sake.

File: snapshots.js [hunk 1]
```diff
@@ -65,14 +65,22 @@ function getExampleUrlsFromExamplePaths(urlPaths) {
...
     1280, // Desktop
   ]);
 
+  if (theme !== 'light') {
+    // Non-light themes are only captured at one width
+    return Array.from(widths);
+  }
+
+  // Light theme is also captured at 375px for mobile
+  widths.add(375);
```
[the remark this answers is not in the document]
Comment (jmuzina) [reply snapshots.js]: @pastelcyborg Good idea, I've introduced a mapping of device/breakpoint name to width in px, and a similar improvement to share the list of all color themes as a constant. Check it out and let me know what you think

File: snapshots.js [hunk 2]
```diff
@@ -106,33 +124,30 @@ async function getPercyConfigURLs() {
   const links = getExampleUrlsFromExamplePaths(await getExampleFiles());
   let urls = [];
 
-  for (let url of links) {
-    const path = new URL(url).pathname.replace(/\/?$/, '/');
-    const widths = await getWidthsForExample(path);
+  for (let link of links) {
+    const path = new URL(link).pathname.replace(/\/?$/, '/');
     // TODO this could be functionalized to get the proper themes for a given example.
     const themes = ['light', 'dark', 'paper'];
```
[the remark this answers is not in the document]
Comment (jmuzina) [reply snapshots.js]: Fixed

**Comment by bartaz** [comment]
> > Just to make sure I understand this: we don't currently take mobile screenshots for dark and paper theme? I didn't realize that; that's not optimal. I sort of understand the logic, but I'd probably suggest we take screenshots of all themes across all breakpoints eventually.
> 
> @pastelcyborg This comes from this [MM message](https://chat.canonical.com/canonical/pl/wh61omg7tibu9nyxy5g6ugwr9w) from @bartaz - it was done as a way of reducing test volume I believe. If we captured in all themes and breakpoints, we'd expect to see test volume increase by around 3x. However, I do agree that it's optimal to snapshot every breakpoint in every theme for thoroughness' sake.

Now that I've been called out directly it feels I need to justify myself :)

Yes, it felt a bit redundant to create all the snapshots in all themes (3 of them) and all screen sizes (2 or 3 of them), building up to 6 or 9 snapshots per example. Especially that they will all change at the same time in a same way.
In our current implementation on theming the only changes between themes are colours, not layout. So it feels unnecessary to create screenshots of medium and screen sizes in all themes, because they are just the same.

That being said, I agree, that it would be better/safer to screenshot everything. But is it worth the increased cost?

Overall, not part of this PR, we can consider this, but I'm afraid it will bring us back to previous snapshot usage (or even above).

File: snapshots.js [hunk 3]
```diff
@@ -106,33 +129,30 @@ async function getPercyConfigURLs() {
...
-    // Take one snapshot per theme at the last breakpoint
-    themes.forEach((theme) => {
+      // Default theme captured responsively, other themes captured at the largest width
       urls.push({
-        url: `${url}?theme=${theme}`,
-        name: `${path.slice(0, path.length - 1)}?theme=${theme}`,
-        widths: [widths[widths.length - 1]],
+        url,
+        name,
+        widths: theme === DEFAULT_COLOR_THEME ? widths : [widths[widths.length - 1]],
```
[the remark this answers is not in the document]
Comment (jmuzina) [reply snapshots.js]: Good catch, that was unneeded and - it's been removed

**Review by bartaz — APPROVED** [review]
LGTM

**Pull request merged**